Update sgp40.py

stansfie opened the pull request:
Hi Nevermore team,

First time writing a github pull, so hope it works!

**Changes to syp40.py**
- Added 'plot_voc' boolean to output only voc vs all parameters
- Added 'voc_scale' float to adjust the scale for voc output for plotting (e.g., 0.5)
- Added INITIALIZE to _init_sgp40
- Removed FEATURE_SET from _init_sgp40 - performs soft reset and results in "unable to write to i2c_device" error from some sgp40 chips
- Removed SOFT_RESET - results in "unable to write to i2c_device" error from some sgp40 chips
- Changed featureset output to 0x3200 from 0x3240 as per Adafruit discussion forums (https://forums.adafruit.com/viewtopic.php?t=202391)

**Humidity correction**
The other neat thing in most printers with organic filters is that they're closed chambers. Relative humidity is therefore a function of change in saturation vapor pressure with temperature with an assumed constant water volume (g/g). With a single chamber thermostat humidity can be corrected assuming a closed chamber.

**Klipper**
Example Klipper config to turn on a nevermore micro in a closed chamber when VOC >120.

_Hardware_
- SGP40 
- Nevermore micro
- Chamber temperature sensor 'chamber'

_Installation_
SGP40 connected to raspberry pi host on i2c.1

printer.cfg:

`## fan - Nevermore Micro with VOC PID control`
`[sgp40]`

`[temperature_sensor chamber]`
`;define your chamber sensor`

`[temperature_fan VOC/2]`
`pin: ; Your Nevemore pin here`
`kick_start_time: 0`
`#shutdown_speed: 0`
`off_below: 0.2`
`max_power: 1.0`
`#fan_speed: 0.6`
`sensor_type: SGP40`
`i2c_mcu: host`
`i2c_bus: i2c.1`
`i2c_speed: 400000`
`voc_scale: 0.5  #set to 50% to enable plotting at a resonable scale`
`ref_temp_sensor: temperature_sensor chamber`
`# ref_humidity_sensor:`
`#min_temp: 0`
`#max_temp: 500`
`control: pid`
`min_temp: 40`
`max_temp: 300`
`#max_delta: 5.0`
`pid_kp: 1`
`pid_ki: 0.2`
`pid_kd: 2.0`
`pid_deriv_time: 5.0`
`min_speed: 0.`
`max_speed: 1.0`
`target_temp: 60 #VOC 120 at voc_scale: 0.5`

**Mainsail output**
<img width="676" alt="Screenshot 2023-11-14 at 8 15 48 pm" src="https://github.com/nevermore3d/Nevermore_Max/assets/25702159/a0bed6fa-e39f-44ea-817c-aabfa8105167">

stansfie committed:
Update sgp40.py

Added 'plot_voc' boolean to output only voc vs all parameters
Added 'voc_scale' float to adjust the scale for voc output for plotting (e.g., 0.5)
Added INITIALIZE to _init_sgp40
Removed FEATURE_SET from _init_sgp40 - performs soft reset and results in "unable to write to i2c_device" error 
Removed SOFT_RESET - results in "unable to write to i2c_device" error 
Changed featureset output to 0x3200 from 0x3240 as per Adafruit discussion forums (https://forums.adafruit.com/viewtopic.php?t=202391)

Added a correction for saturation vapor pressure with changing temperature to adjust humidity assuming initial humidity 50% with temperature 25C

SanaaHamel commented on Software/Klipper/sgp40.py:
```diff
@@ -84,19 +88,24 @@ def get_report_time_delta(self):
-        self_test = self._read_and_check(SGP40_CMD["SELF_TEST"], wait_time_s = 0.5)
-        if self_test[0] != 0xD400:
-            logging.exception("sgp40: Self test error")
+        # self_test = self._read_and_check(SGP40_CMD["SELF_TEST"], wait_time_s = 0.5)
+        # if self_test[0] != 0xD400:
+        #    logging.exception("sgp40: Self test error")
```
The failure you described is an I2C comm issue. It doesn't make sense to plaster over the problem by disabling the self-test.

stansfie replied:
Oops. This is back

SanaaHamel commented on Software/Klipper/sgp40.py:
```diff
@@ -84,19 +88,24 @@ def get_report_time_delta(self):
         # Reset
-        self._read_and_check(SGP40_CMD["SOFT_RESET"], read_len = 0, wait_time_s = 0.5)
-
+        # self._read_and_check(SGP40_CMD["SOFT_RESET"], read_len = 0, wait_time_s = 0.5)
+    
```
Just remove it entirely.
1) If I interpret the data sheet correctly this is a general call reset so everyone on the bus is supposed to restart, which is extremely rude to other systems.
2) Wouldn't work anyways. General calls are issued to address 0x00. This issues to the SGP40 address and sends a nonsense opcode.

stansfie replied:
Done

SanaaHamel commented on Software/Klipper/sgp40.py:
```diff
@@ -111,16 +120,23 @@ def _sample_sgp40(self, eventtime):
 
         measured_time = self.reactor.monotonic()
-        self._callback(self.mcu.estimated_print_time(measured_time), self.temp)
+        self._callback(self.mcu.estimated_print_time(measured_time), self.voc * self.voc_scale)
```
Why scale the callback but not the `get_status` result?

stansfie replied:
Great point - this is also done

SanaaHamel commented on Software/Klipper/sgp40.py:
```diff
@@ -84,19 +88,24 @@ def get_report_time_delta(self):
 
     def _init_sgp40(self):
+        # Check serial number
+        serial_number = self._read_and_check(SGP40_CMD["INITIALIZE"], wait_time_s = 0.5)
+        if serial_number[0] != 0x0000:
+            logging.exception("sgp40: Serial number does not match")
```
I have no idea what you're trying to do here.
1) The serial # op is pure, so IDK why bother if you're not going to log the serial # or something
2) Why exclude higher serial #s?

stansfie replied:
I've removed the call to get the serial number; it's not required and I initially saw it in a sensirion initialization routine. The only call to sgp40 that's really needed in the _init_sgp40 is the self_test which you pointed out above.

SanaaHamel reviewed (changes requested):
tl;dr: Some good, some bad.

Papering over I2C comm issues by commenting out checks can work for your machine, but isn't an acceptable general solution.

SanaaHamel commented on Software/Klipper/sgp40.py:
```diff
@@ -14,6 +15,7 @@
 SGP40_CMD = {
+    "INITIALIZE" : [0x36, 0x82],
     "SOFT_RESET" : [0x00, 0x06],
     "FEATURE_SET" : [0x20, 0x2F],
```
`0x202F` was an op-code from the SGP30 device, and isn't officially supported by the SGP40.
Why was this ever used?

stansfie replied:
Great point! It's gone.

stansfie committed:
Update sgp40.py

Removed FEATURE_SET and related callbacks as not included in sgp40 
Removed SOFT_RESET from _init

Changed INITIALIZE to GET_SERIAL which makes more sense and removed it from the _init as it's not required.

added voc_scale to get_status output

stansfie commented on Software/Klipper/sgp40.py:
```diff
@@ -84,19 +88,24 @@ def get_report_time_delta(self):
-        if feature_set[0] != 0x3220:
-            logging.exception("sgp40: wrong feature set returned by sensor")
+        # feature_set = self._read_and_check(SGP40_CMD["FEATURE_SET"], wait_time_s = 0.5)
+        # if feature_set[0] != 0x3200:
```
Done

stansfie commented on Software/Klipper/sgp40.py:
```diff
@@ -14,6 +15,7 @@
 SGP40_WORD_LEN = 2
 
 SGP40_CMD = {
+    "INITIALIZE" : [0x36, 0x82],
```
Renamed INITIALIZE to GET_SERIAL, but removed calls to get a serial number as it's not strictly needed

stansfie commented:
Thanks so much for your suggestions @SanaaHamel! Corrections have been made to the sgp40.py to address all of the above. Hope it's all more consistent now.

SanaaHamel commented on Software/Klipper/sgp40.py:
```diff
@@ -88,24 +87,12 @@ def get_report_time_delta(self):
-    
+        self_test = self._read_and_check(SGP40_CMD["SELF_TEST"], wait_time_s = 0.5)
+        if self_test[0] != 0xD400:
+           logging.exception("sgp40: Self test error")
```
This should be `logging.error` instead of `logging.exception`, `exception` attaches exception info and should only be used in a `except` block's scope,

SanaaHamel commented on Software/Klipper/sgp40.py:
```diff
@@ -194,7 +181,7 @@ def _generate_crc(self, data):
         # HACKL can only plot on mainsail/fluidd if VOC index is a temperature
         if self.plot_voc:
-            return {"temperature": self.voc }
+            return {"temperature": self.voc * self.voc_scale }
```
For consistency, also scale the VOC value in the non-plot case?

stansfie replied:
Done - VOC output is now scaled consistently independent of output type (plot_voc = TRUE | FALSE)

TL;DR I've left the sensor output as "voc"

Re: renaming output variables. I guess this is a question of accuracy, compatibility and semantics. Considering gas sensors are becoming more capable with multiple outputs (e.g., sgp30 - CO2 and VOC) specifying the input required for Nevermore fan/scrubber control algorithm may be the more future proof option rather than renaming the output of sensor modules to 'gas'.

Consider a hypothetical sensor that could output formaldehyde, VOC and ultrafine particle measures - VOC output could be used to control activated carbon scrubbing, formaldehyde output would require catalyst breakdown and ultrafine particle output would indicate the need for HEPA filtration. Maintaining accurate nomenclature here is useful in determining what needs to be run.

I think it makes sense to keep sensor output as specific as possible and allow flexible input options for the nevermore control algorithm.

SanaaHamel reviewed (changes requested):
Two last nitpicks, but I think it should be good to go afterwards.

stansfie committed:
Update sgp40.py

updated `logging.exception` to `logging.error` for _init_sgp40

consistently scaled voc output by voc_scale

SanaaHamel reviewed (approved):
LGTM
